src/features/terminal/utils.ts

```diff
@@ -1,19 +1,115 @@
 import * as path from 'path';
-import { Terminal, TerminalOptions, Uri } from 'vscode';
+import { Disposable, Terminal, TerminalOptions, Uri } from 'vscode';
 import { PythonEnvironment, PythonProject, PythonProjectEnvironmentApi, PythonProjectGetterApi } from '../../api';
-import { sleep } from '../../common/utils/asyncUtils';
+import { timeout } from '../../common/utils/asyncUtils';
+import { createSimpleDebounce } from '../../common/utils/debounce';
+import { onDidChangeTerminalShellIntegration, onDidWriteTerminalData } from '../../common/window.apis';
 import { getConfiguration, getWorkspaceFolders } from '../../common/workspace.apis';
 
 export const SHELL_INTEGRATION_TIMEOUT = 500; // 0.5 seconds
 export const SHELL_INTEGRATION_POLL_INTERVAL = 20; // 0.02 seconds
 
+/**
+ * Three conditions in a Promise.race:
+ * 1. Timeout based on VS Code's terminal.integrated.shellIntegration.timeout setting
+ * 2. Shell integration becoming available (window.onDidChangeTerminalShellIntegration event)
+ * 3. Detection of common prompt patterns in terminal output
+ */
 export async function waitForShellIntegration(terminal: Terminal): Promise<boolean> {
-    let timeout = 0;
-    while (!terminal.shellIntegration && timeout < SHELL_INTEGRATION_TIMEOUT) {
-        await sleep(SHELL_INTEGRATION_POLL_INTERVAL);
-        timeout += SHELL_INTEGRATION_POLL_INTERVAL;
+    if (terminal.shellIntegration) {
+        return true;
+    }
+
+    const config = getConfiguration('terminal.integrated');
+    const timeoutValue = config.get<number | undefined>('shellIntegration.timeout');
+    const timeoutMs = timeoutValue === undefined || -1 ? 5000 : timeoutValue;
```

Comment:
You can make it more consistent with `getShellIntegrationTimeout` by checking `terminal.integrated.shellIntegration.enabled` (mostly will match `siInjectionEnabled`) and the window's `remoteAuthority` for isRemote.

Reply:
Sounds good,, I think the remoteAuthority is from proposed api though: https://github.com/microsoft/vscode/blob/c23478af52b09c278e21e8396a0e120eb134ae3a/src/vscode-dts/vscode.proposed.resolvers.d.ts#L472  

Would `env.remoteName` be good enough? [28f5261](https://github.com/microsoft/vscode-python-environments/pull/952/commits/28f5261d294977bc53cd252d01b478db291914cc)